cmake: default to linking 'app' with the interface library 'FS'

SebastianBoe opened the pull request:
This patch removes the need for application build script code to
explicitly link 'app' with a filesystem implementation.

It does this by introducing a zephyr interface library called 'FS'
that contains the usage requirements for linking with the filesystem
library subsys__fs and using Kconfig to default to linking the 'app'
library with this interface library.

SebastianBoe committed:
cmake: default to linking 'app' with the interface library 'FS'

This patch removes the need for application build script code to
explicitly link 'app' with a filesystem implementation.

It does this by introducing a zephyr interface library called 'FS'
that contains the usage requirements for linking with the filesystem
library subsys__fs and using Kconfig to default to linking the 'app'
library with this interface library.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
cmake: Remove duplicate invocations of target_link_libraries on app

It is not necessary to link 'app' with ELMFAT or NFFS because their
usage requirements are covered by the 'APP_LINK_WITH_MBEDTLS'
mechanism that automatically links 'app' with the interface library
FS.

This patch removes the redundant target_link_libraries invocations.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

nashif commented on subsys/fs/Kconfig:
```diff
@@ -33,6 +33,14 @@ config SYS_LOG_FS_LEVEL
 
 if FILE_SYSTEM
 
+config APP_LINK_WITH_FS
```
I do not see this being used

SebastianBoe replied:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/app/boilerplate.cmake#L318

pboettch commented on subsys/fs/CMakeLists.txt:
```diff
@@ -8,8 +9,10 @@ if(CONFIG_FILE_SYSTEM)
+  zephyr_library_link_libraries(FS)
+
+  target_link_libraries_ifdef(CONFIG_FAT_FILESYSTEM_ELM FS INTERFACE ELMFAT)
+  target_link_libraries_ifdef(CONFIG_FILE_SYSTEM_NFFS   FS INTERFACE NFFS)
```
Shouldn't these two lines be in the CMakeList.txt of NFFS and ELMFAT respectively?

SebastianBoe replied:
Maybe, but they can't because FS can only be modified in the CMakeLists.txt directory that created it.

ifillonneau commented:
in subsys/settings/src/CMakeLists.txt, is it necessary to keep following line:
zephyr_include_directories($ENV{ZEPHYR_BASE}/ext/fs/nffs/include)

SebastianBoe committed:
cmake: settings: Don't add ext nffs include dir to global includes

The settings subsystem has been adding nffs's include dir to the
global set of paths. Presumably because app's will need acces. But
this is no longer necessary as we default to linking 'app' with FS,
which again has the NFFS include paths.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

nashif commented:
recheck
